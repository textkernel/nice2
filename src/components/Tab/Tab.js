import React from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import styles from './Tab.scss';

const { block } = bem({
    name: 'Tab',
    classnames: styles,
    propsToMods: []
});

const Tab = props => {
    const { children, href, id, label, ...rest } = props;

    return (
        <div {...rest} {...block(props)}>
            {children}
        </div>
    );
};

Tab.propTypes = {
    /** The tab content */
    children: PropTypes.node.isRequired,
    /** URL / route for this tab */
    href: PropTypes.string,
    /** Unique ID for this tab */
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    /** Label for this tab */
    label: PropTypes.node.isRequired
};

Tab.defaultProps = {
    href: '#'
};

export default Tab;
