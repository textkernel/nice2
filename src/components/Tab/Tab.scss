.Tab {
    background-color: var(--color-background);
    padding: var(--spacing-3x);
}
