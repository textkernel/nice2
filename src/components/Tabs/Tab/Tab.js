import PropTypes from 'prop-types';

const Tab = ({ children }) => children;

Tab.displayName = 'Tab';

Tab.propTypes = {
    /** The tab content */
    children: PropTypes.node.isRequired,
    /** Unique ID for this tab */
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    /** Label for this tab */
    label: PropTypes.node.isRequired,
};

Tab.defaultProps = {};

export default Tab;
