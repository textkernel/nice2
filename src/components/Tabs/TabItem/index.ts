export { default as TabItem } from './TabItem';
export { Props as TabItemProps } from './TabItem';
