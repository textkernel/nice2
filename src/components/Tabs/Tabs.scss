%tab {
    background-color: var(--color-neutral-25);
    border: {
        color: var(--color-neutral);
        style: solid;
        top-left-radius: var(--border-radius);
        top-right-radius: var(--border-radius);
        width: var(--line-normal);
        bottom-width: 0;
    }
    color: var(--link-color-normal);
    cursor: pointer;
    outline: none;
    padding: var(--spacing-2x) var(--spacing-4x);
    line-height: var(--line-height-normal);
    text-decoration: none;
    transition: border-color var(--transition-duration);

    + .Tabs__tab,
    + .Tabs__tabActive {
        margin-left: var(--spacing-normal);
    }

    &:hover {
        border-color: var(--color-neutral-60);
    }
}

.Tabs {
    color: black;
    font: {
        family: var(--font-family-primary);
        size: var(--font-size-normal);
    }

    &__container {
        border-bottom: {
            color: var(--color-neutral);
            style: solid;
            width: var(--line-normal);
        }
        display: flex;
        justify-content: flex-start;
        padding: {
            left: var(--spacing-3x);
            right: var(--spacing-3x);
        }
    }
    &__tab {
        @extend %tab;
    }
    &__tabActive {
        @extend %tab;
        background-color: var(--color-background);
        color: var(--color-foreground);
        margin-bottom: -1px;

        &:hover {
            border-color: var(--color-neutral);
        }
    }
}
