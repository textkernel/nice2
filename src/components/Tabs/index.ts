export { TabItem } from './TabItem';
export { TabsBar } from './TabsBar';
