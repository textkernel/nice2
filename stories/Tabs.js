import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, select, withKnobs } from '@storybook/addon-knobs';
import { Tab, Tabs, TabContent, TabItem, TabMenu } from '@textkernel/oneui';

const tabs = [
    {
        id: 'first',
        label: 'First tab',
        content: 'Some content',
    },
    {
        id: 'second',
        label: 'Second tab',
        content: 'Some more content',
    },
    {
        id: 'third',
        label: 'Third tab',
        content: 'Even more content',
    },
];

const tabIds = tabs.map(tab => tab.id);

storiesOf('Tabs', module)
    .addDecorator(withKnobs)
    .add('Simple (managed) tabs', () => (
        <Tabs
            activeTabId={select('Active tab', tabIds, tabIds[0])}
            gutters={boolean('Show gutters', false)}
            onChange={tabId => {
                console.log(`Switched to ${tabId} tab`);
            }}
        >
            {tabs.map(tab => (
                <Tab label={tab.label} id={tab.id} key={tab.id}>
                    {tab.content}
                </Tab>
            ))}
        </Tabs>
    ))
    .add('Controlled mode', () => (
        <div>
            <TabMenu
                activeTabId="second"
                onChange={(e, id) => {
                    e.preventDefault();
                    console.log(`Switch to ${id} tab requested`);
                }}
                gutters={boolean('Show gutters', false)}
            >
                <TabItem id="first" label="Some tab" />
                <TabItem id="second" label="Another tab" />
            </TabMenu>
            <div style={{ padding: 9 }}>Some other content in between tab menu and tab content</div>
            <TabContent>Some content</TabContent>
        </div>
    ));
