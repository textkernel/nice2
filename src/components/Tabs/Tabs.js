import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import Tab from '../Tab';
import styles from './Tabs.scss';

class Tabs extends PureComponent {
    constructor(props) {
        super(props);

        const { activeTabId } = props;

        this.state = {
            activeTabId
        };
    }

    render() {
        const { activeTabId: idFromProps, children, ...rest } = this.props;
        const { activeTabId } = this.state;

        return (
            <div {...rest} {...this.block()}>
                <div {...this.elem('container')}>
                    {React.Children.map(children, tab => {
                        const active = activeTabId === tab.props.id;
                        return (
                            <a
                                href="#" {...this.elem(active ? 'tabActive' : 'tab')}
                                onClick={ e => {
                                    e.preventDefault();
                                    console.log(tab.props.id);
                                    this.setState({
                                        activeTabId: tab.props.id
                                    }, () => {
                                        console.log(this.state.activeTabId);
                                    });
                                }}
                            >
                                {tab.props.label}
                            </a>
                        );
                    })}
                </div>
                {React.Children.map(children, tab => {
                    if (tab.props.id !== activeTabId) {
                        return null;
                    }
                    return tab;
                })}
            </div>
        );
    }
};

Tabs.propTypes = {
    /** ID of currently active tab */
    activeTabId: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number
    ]).isRequired,
    /** The tabs */
    children: PropTypes.oneOfType([
        PropTypes.instanceOf(Tab),
        PropTypes.arrayOf(PropTypes.instanceOf(Tab))
    ]).isRequired
};

Tabs.defaultProps = {};

export default bem(styles)(Tabs);
