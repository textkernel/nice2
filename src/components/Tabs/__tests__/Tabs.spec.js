import React from 'react';
import toJson from 'enzyme-to-json';
import Tabs from '../Tabs';
import Tab from '../Tab';

describe('<Tabs> that renders tab container with some tabs', () => {
    it('should render simple tabs (without onChange callback prop)', () => {
        const onChange = jest.fn();

        const wrapper = mount(
            <Tabs activeTabId="1">
                <Tab id="1" label="First">
                    Some content
                </Tab>
                <Tab id="2" label="Second">
                    Some more content
                </Tab>
            </Tabs>
        );
        expect(toJson(wrapper)).toMatchSnapshot();

        wrapper
            .find('TabMenu')
            .find('TabItem')
            .at(1)
            .simulate('click');

        expect(onChange).not.toHaveBeenCalled();
    });

    it('should switch tabs correctly', () => {
        const onChange = jest.fn();

        const wrapper = mount(
            <Tabs activeTabId="1" onChange={onChange}>
                <Tab id="1" label="First">
                    Some content
                </Tab>
                <Tab id="2" label="Second">
                    Some more content
                </Tab>
            </Tabs>
        );
        wrapper
            .find('TabMenu')
            .find('TabItem')
            .at(1)
            .simulate('click');
        expect(wrapper.state().activeTabId).toBe('2');
        expect(onChange).toHaveBeenCalledWith('2');
    });

    it('should switch tabs correctly (programmatically)', () => {
        const wrapper = mount(
            <Tabs activeTabId="1">
                <Tab id="1" label="First">
                    Some content
                </Tab>
                <Tab id="2" label="Second">
                    Some more content
                </Tab>
            </Tabs>
        );
        wrapper.setProps(
            {
                activeTabId: '2'
            },
            () => {
                expect(wrapper.state().activeTabId).toBe('2');
                expect(wrapper.state().derivedTabId).toBe('2');
                expect(wrapper.find('TabContent').text()).toEqual('Some more content');
            }
        );
    });
});
