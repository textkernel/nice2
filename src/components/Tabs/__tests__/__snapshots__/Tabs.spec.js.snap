// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs> that renders tab container with some tabs should render simple tabs (without onChange callback prop) 1`] = `
<Tabs
  activeTabId="1"
  gutters={false}
  onChange={null}
>
  <TabMenu
    activeTabId="1"
    gutters={false}
    onChange={[Function]}
  >
    <div
      className="TabMenu"
      role="tablist"
    >
      <TabItem
        id="1"
        isActive={true}
        key=".$.0"
        label="First"
        onClick={[Function]}
      >
        <a
          aria-selected={true}
          className="TabItem TabItem--isActive"
          href="#"
          id="1"
          onClick={[Function]}
          role="tab"
          tabIndex={0}
        >
          First
        </a>
      </TabItem>
      <TabItem
        id="2"
        isActive={false}
        key=".$.1"
        label="Second"
        onClick={[Function]}
      >
        <a
          aria-selected={false}
          className="TabItem"
          href="#"
          id="2"
          onClick={[Function]}
          role="tab"
          tabIndex={0}
        >
          Second
        </a>
      </TabItem>
    </div>
  </TabMenu>
  <TabContent
    aria-labelledby="1"
    key=".0"
  >
    <div
      aria-labelledby="1"
      className="TabContent"
      role="tabpanel"
    >
      Some content
    </div>
  </TabContent>
</Tabs>
`;
